lib/Doctrine/ORM/Internal/Hydration/ScalarColumnHydrator.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Internal\Hydration;
+
+/**
+ * Hydrator that produces one-dimensional array "graph" that is often (not always) interchangeable
+ * with the corresponding object graph for read-only access.
+ */
+class ScalarColumnHydrator extends AbstractHydrator
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function hydrateAllData()
+    {
+        if (!isset($this->_rsm->indexByMap['scalars'])) {
+            return $this->_stmt->fetchOne();
+        }
+
+        $result = $this->hydrateAssociativeData();
+
+        if (empty($result)) {
+            return $result;
+        }
+
+        $indexColumn = $this->_rsm->scalarMappings[$this->_rsm->indexByMap['scalars']];
+
+        $keys = array_keys(reset($result));
+
+        $column = isset($keys[1]) && $keys[0] === $indexColumn ? $keys[1] : $keys[0];
+
+        return array_column($result, $column, $indexColumn);
+    }
+
+    /**
+     * @throws \Doctrine\DBAL\Driver\Exception
+     * @return array
+     */
+    protected function hydrateAssociativeData()
+    {
+        $result = [];
+
+        while ($data = $this->_stmt->fetchAssociative()) {
```

Comment:
Shouldn't `fetchOne` be used here too?

Reply:
https://github.com/doctrine/orm/pull/8919#discussion_r693490824

`fetchOne` throws error if we use `gatherScalarRowData` this to hydrate values

Reply:
It goes into Infinite loop

Reply:
Not sure `gatherScalarRowData` will be useful here :thinking: I mean we don't care about getting the fieldname that corresponds to the column name we are fetching here, since in the end we will not return it.

Reply:
Updated Hydrator, now it's simplified.